### DISCUSSION TITLE
NFC Fails on libre2 direct for ~ hour - smoothed data problem?

### DISCUSSION THREAD

**Discussion by nigelkeirross** [NightscoutFoundation/xDrip#3527] [Q&A]
I am running Libre2 direct (no bridge). I had been receiving reading happily for some time, but then BT became disconnected. After a missed reading, I attempted an NFC scan, but this failed to generate a reading, despite making the usual NFC vibrations & sounds.
I looked in the log file, as seen below & I noticed that it appeared to be getting a correct/reasonable raw reading, but the SMOOTHED readings were 0, or in one case -1, and this seems to be the cause of the reading being rejected.

Reading Data
18:07 Removing point glucoseData = {sensorTime = 14809 glucoseLevel=113 glucoseLevelRaw=1486 glucoseLevelRawSmoothed = 0 flags=2080 source=FRAM glucoseLevelSmoothed-1}
and
LibreTrendPoint
18:07 LibreTrendPoint.isError returning true {sensorTime=14809 rawSensorValue=1486 flage=2080 source=FRAM glucoseLevel 0}

Can anyone shed any light on why this might be happening, and what can be done to fix this? It did eventually recover, but I missed about an hour of readings.
![NFC-fails](https://github.com/NightscoutFoundation/xDrip/assets/21008320/235ba4df-b13a-4c07-bef5-f09934e4aa08)

**Comment by PhilTU** [comment]
I've seen this often, values from scanning are deleted.

I think the root cause is zero values being assigned when an error is flagged on data read from FRAM. The zero then leads to an out of bounds smoothed value which is rejected. The one minute values at 14806 and 14809 are flagged in your screenshot.

Sometimes OOP2 shows a negative value. It appears to happen most if you NFC scan around the time of a Bluetooth transmission, if yo usee OOP2 collect a value and wait 20s then scan it may occur less (though often we scan because there is no BLE value).

**Reply by nigelkeirross** [reply]
Thanks.  is FRAM the sensor memory? Thre NFC failed again this morning for about 45 minutes - and same pattern - it seemingly gets good readings, but according to the logs, the smoothed value is 0, or GlucoseLevelSmoother -1. This to me points to a bug in the smoothing process. In this case 3 consecutive glucose readings at 2743,2744 & 2745 all Glucoselevel 101, GlucoseLevelRaw 1075. So it looks like xDrip is getting good readings from the NFC, BUT somehow the smoothed values become 0, and so no reading accepted? Is there a way to turn smoothing off?

**Reply by PhilTU** [reply]
> Thanks. is FRAM the sensor memory

Yes, the Flash RAM that NFC reads.

If you scroll down the values you might see the erroneous ones in the 15 x 1 minute values (rather than the 15 minute averages which are more stable)

**Comment by PhilTU** [comment]
Were there any flags setting an error condition?

**Comment by nigelkeirross** [comment]
I don't really understand the question. Where would the flags be set, and
what would they look like?

**Reply by PhilTU** [reply]
they're in your originl image - bad data values flagged by sensor perhaps 
![image](https://github.com/NightscoutFoundation/xDrip/assets/76240903/ea40c579-db08-4206-8c61-c4dd7173e15f)

**Reply by nigelkeirross** [reply]
Thanks for the explanation. Yes, in the LibreTrendPoint section, both timestamp 2743&2745 had flags 2080. In ReadingData section 2743&2745 have flags2080, but 2744 has flags=0.
Where do I find how to decode the flags? 2080 seems common. Where are they documented?

**Reply by PhilTU** [reply]
I don't know if they are, it's a sensor error code I suspect and we're using software that is effectively reverse engineered https://github.com/captainbeeheart/openfreestyle/blob/master/docs/reverse.md . You might get a "try again in 10 minutes" alert from Librelink in this scenario.

I don't fully understand it, but my feeling is the flagged values are set to zero by LibreTrendPoint for safety. Why the smoothed value is -1 I don't know, it must be a response to the 0 value being below the 40 mg/dl sensor minimum ?

One day I might dig into the code and try to understand how this works. You raise a good point though, the negative values and missing data that occur with NFC scans are a pain, but do we need to NFC scan much these days ?

**Comment by nigelkeirross** [comment]
== but do we need to NFC scan much these days ?

Thanks. I only NFC when I get a missed reading. It also seems to (re)
initiate a connection. I sometimes have periods where I lose connections
for up to 2hours at a time so end up NFCing every 5 minutes for hours at a
time.

**Reply by PhilTU** [reply]
It does initiate a connection but every 5 mins might be too much, the sensor broadcasts every 2 minutes when it has lost connection and you may find it takes several minutes to reconnect.

It's a shame there isn't more visibility of connection activity, the BT status page shows a countdown.

**Comment by gui-dos** [comment]


**Reply by PhilTU** [reply]
> ```java
> RF(32)
> ```

Thanks, the RF code came up earlier on xDrip from a sensor that was running very low glucose readings.

**Comment by nigelkeirross** [comment]
Thanks. From memory (I am travelling), my flags were 2080, so not sure how
that relates. What does RF mean?
Note I never knowingly NFC when BT is connected.

**Reply by nigelkeirross** [reply]
Thanks. So RF in this context implies NFC (being short for RFID), rather than RF in its more generic sense, since BT is also a RF transmission.?

**Comment by nigelkeirross** [comment]
Guido,
In general I only NFC when connection is lost and I have missed a (5
minute) reading. Is it easy to see from the logs which readings come from
NFC Vs BT. (I can look in oop2 to find recent BT readings).
Also where do I find the decompiled code you refer to?

**Reply by PhilTU** [reply]
> Is it easy to see from the logs which readings come from NFC Vs BT.

Source = BLE vs Source = FRAM. In the yellow box is a standard 1 minute value received by BLE

The yellow circles are around flagged errors where BLE was the source :-

![image](https://github.com/NightscoutFoundation/xDrip/assets/76240903/59be486c-a20f-406b-8c2d-d1ee83ecdbfa)

**Reply by nigelkeirross** [reply]
Thank-you that is very helpful. So FRAM is NFC, and BLE is a normal reading from a BT connection. 
The reading at 17325 is interesting - the raw value 761 is similar to the one 10 mins later at 17345 789, so plausible,and yet the smoothed values for this are set to 0. & -1. Again this seems to me to point to the smoothing function often returning 0 where perhaps it shouldn't.

**Reply by PhilTU** [reply]
It's one of those things that could take a whole day of debate in a meeting - what value to assign to values flagged as error state and how to handle them in the smoothing algorithm. Rather than setting as zero I would personally treat a flagged value as not received.

**Reply by PhilTU** [reply]
> Every minute the Libre 2 sends 10 measurements (4 bytes each): seven “trend” values and three “historical” values.

thanks for that. Interesting it sends every other one-minute trend value. I suppose a minute later the gap is filled so it's an interesting way of covering missed transmissions.

**Comment by gui-dos** [comment]


**Reply by nigelkeirross** [reply]
Thanks. So 2080 = 2148 (Thermistor) || RF (32). How do I know to interpret this as an error? Isn't "Thermistor out of range" also an error?

**Reply by nigelkeirross** [reply]
@gui-dos Thanks for the explanation of the jadx tool & looking into the Abbot APK. I have never done that but will try.

**Reply by nigelkeirross** [reply]
@PhilTU  Following your suggestion, I updated my java code base to April 2026. The problem of deleting non-zero raw values persisted, so with the help og Gemini I have added a setting such that I now by default prevent deletions of all non-zero values then flags have been set. These flags are set by the sensor as a warning, but it seems that xDrip is very strict in deleting any reading with flags set. Almost without exception, the values are "resonable" and in line with earlier and later ones. This has helped keep the flow of readings arriving. However, recently with a new sensor I saw a reading being rejected for no apparent reading. Gemini analysed the log files, and diagnosed the "NFC Shdow" effect. This reminded me of your earlier post: "I've seen this often, values from scanning are deleted. I think the root cause is zero values being assigned when an error is flagged on data read from FRAM. The zero then leads to an out of bounds smoothed value which is rejected." 
I had these logs:
2 ^ 2026-05-26 10:34:34 ^ Sentinam ^ 1-min BG @  18720 mmol=7.2 (983)  (59s)
3 ^ 2026-05-26 10:33:35 ^ jamorham librereceiver ^ Sensor age has gone backwards!!! nfc_sa=18718 vs last in trend=18717
2 ^ 2026-05-26 10:33:35 ^ ReadingData ^ Removing point glucoseData =  { sensorTime = 18719 glucoseLevel = 97 glucoseLevelRaw = 988 glucoseLevelRawSmoothed = 0 flags = 1 source = BLE  glucoseLevel 97 glucoseLevelSmoothed -1}
2 ^ 2026-05-26 10:33:34 ^ Sentinam ^ 1-min BG @  18719 mmol=7.2 (988)  (58s)
__________-

**Reply by nigelkeirross** [reply]
@philtu This is one of the most fascinating edge-case bugs in the entire xDrip architecture. You have captured a perfect chronological log of the system tripping over its own safety mechanisms.
To answer your question directly: Point 18719 was assassinated by a "4-Minute Shadow of Death" caused by an NFC scan you performed one minute earlier.
Here is the exact step-by-step mathematical autopsy of what happened between 10:32 and 10:34, followed by the surgical fix.

The Autopsy of 18719
1. The NFC Scan (10:32)
Around 10:32, you scanned the sensor with NFC. We know this because the log at 10:33 says nfc_sa=18718 (NFC Sensor Age = 18718).
When you scan via NFC, xDrip reads the history and saves the points to the global database with the source tag FRAM.

2. The Flag Trap
At the time of your scan, the sensor was outputting a physical flag of 1 (which is normal for Libre 2 Plus variants during certain physiological states).
If you look inside LibreTrendPoint.java, there is a very strict legacy rule for NFC (FRAM) data:

Java
if (source == GlucoseData.DataSource.FRAM) {
    ret |= (flags != 800) && (flags != 0); // If flag is not 800 or 0, MARK AS ERROR!
}
Because the flag was 1, xDrip looked at the NFC point at 18718 and completely falsely declared: "This point is corrupted hardware data."

3. The Shadow of Death (10:33)
xDrip has a safety protocol in ReadingData.java called calculateErrorSet. If it believes a point is corrupted, it assumes the sensor is physically failing and casts a 4-minute shadow of doubt into the future (ERROR_INFLUENCE = 4).
Because 18718 was marked as an error, xDrip added 18718, 18719, 18720, and 18721 to the errorHash blacklist.

When the BLE packet for 18719 arrived at 10:33, it fell perfectly into this blacklist. Your 1-minute pipeline dutifully printed 1-min BG @ 18719, but a microsecond later, ClearErrors() saw it in the blacklist and ruthlessly purged it.

4. The Backwards Time Illusion (10:33)
Because 18719 was purged from the array, the highest surviving point in the array was 18717.
The LibreAlarmReceiver then compared the surviving array against the NFC scan from a minute ago: "Wait, the NFC age is 18718, but the newest array point is 18717? Sensor age has gone backwards!!!" (This explains your mysterious jamorham log!)

5. The Resurrection (10:34)
Why didn't 18720 get killed by the shadow at 10:34?
Because Abbott's BLE payloads contain overlapping history (0, -2, -4 minutes). The BLE packet at 10:34 contained data for 18720, but it ALSO contained redundant BLE data for 18718.
When xDrip parsed this BLE packet, it overwrote the old FRAM point at 18718 with a new BLE point.
If you look back at the isError() logic, BLE points ignore flags. Suddenly, 18718 was no longer considered an error, the 4-minute shadow instantly vanished, and 18720 was allowed to live!

**Reply by nigelkeirross** [reply]
@PhilTU This was the proposed fix from Gemini, which has done the trick..
The Fix
To stop NFC scans from generating these 4-minute blackout shadows on your newer sensors, we need to update the legacy flag rules in LibreTrendPoint.java to respect the Sentinam variant memory we built.

Open dexdrip/utils/LibreTrendPoint.java and replace your existing isError() function with this updated version:

Java
    public boolean isError() {
        boolean ret = rawSensorValue == 0;
        if (source == GlucoseData.DataSource.FRAM) {
            // NKR Gemini - Relax strict NFC flag rules for Libre 2 Plus variants
            boolean isPlus = com.eveningoutpost.dexdrip.utilitymodels.Pref.getBoolean("sentinam_libre2_plus", false);
            if (!isPlus) {
                // Classic Libre 1/2 rule: mark as error if flag is not 800 or 0
                ret |= (flags != 800) && (flags != 0);
            }
        } else if (source == GlucoseData.DataSource.BLE) {
            // Condition already checked (rawSensorValue == 0)
        } else {
            ret = false;
        }
        if(ret) {
            Log.e(TAG, "LibreTrendPoint.isError returning " + ret + " " + toString());
        }
        return ret;
    }
Once compiled, you can scan your sensor with NFC as many times as you want without triggering the "Grim Reaper," and your BLE 1-minute graph will remain completely uninterrupted!

**Reply by PhilTU** [reply]
Excellent. I was sure it had to be something wrong that was filling the logs with deletion of data that was perfectly valid.
